api/types/role.go

```diff
@@ -463,39 +466,79 @@ func (r *RoleV6) GetKubeResources(rct RoleConditionType) []KubernetesResource {
 	if rct == Allow {
 		return r.convertAllowKubernetesResourcesBetweenRoleVersions(r.Spec.Allow.KubernetesResources)
 	}
-	return r.convertDenyKubernetesResourcesBetweenRoleVersions(r.Spec.Deny.KubernetesResources)
+	return r.convertKubernetesResourcesBetweenRoleVersions(r.Spec.Deny.KubernetesResources)
 }
 
-// convertDenyKubernetesResourcesBetweenRoleVersions converts Kubernetes resources between role versions.
+// convertKubernetesResourcesBetweenRoleVersions converts Kubernetes resources between role versions.
 // This is required to keep compatibility between role versions to avoid breaking changes
 // when using an older role version.
 //
 // For roles v8, it returns the list as it is.
 //
 // For roles <=v7, it maps the legacy teleport Kinds to k8s plurals and sets the APIGroup to wildcard.
-func (r *RoleV6) convertDenyKubernetesResourcesBetweenRoleVersions(resources []KubernetesResource) []KubernetesResource {
+func (r *RoleV6) convertKubernetesResourcesBetweenRoleVersions(resources []KubernetesResource) []KubernetesResource {
 	switch r.Version {
 	case V8:
 		return resources
 	default:
 		v7resources := slices.Clone(resources)
+		var extraResources []KubernetesResource
 		for i, r := range v7resources {
+			// "namespace" kind used to mean "namespaces" and all resources in the namespace.
+			// It is now represented by 'namespaces' for the resource itself and wildcard for
+			// all resources in the namespace.
+			if r.Kind == KindKubeNamespace {
+				r.Kind = Wildcard
+				r.Namespace = r.Name
+				r.Name = Wildcard
+				r.APIGroup = Wildcard
+				v7resources[i] = r
+				extraResources = append(extraResources, KubernetesResource{
+					Kind:  "namespaces",
+					Name:  r.Namespace,
+					Verbs: r.Verbs,
+				})
+				continue
+			}
+			// The namespace field was ignored in v7 for global resources.
+			if r.Namespace != "" && slices.Contains(KubernetesClusterWideResourceKinds, r.Kind) {
+				r.Namespace = ""
+			}
 			if k, ok := KubernetesResourcesKindsPlurals[r.Kind]; ok { // Can be empty if the kind is a wildcard.
 				r.Kind = k
 			}
 			r.APIGroup = Wildcard
```

Comment:
Here I would specify the API group based on the legacy value to avoid weird behavior if two different apis objects are named `Secret` while you only want to give access to the core api. Can be done in a separate PR

Reply:
Done in a different PR https://github.com/gravitational/teleport/pull/54891/files#diff-1534c9aa06df308bf26ce16a5c1a5eb7cb06b45462370fa0e71804f545e6094d